server/src/logic/services/auth.service.ts

```diff
@@ -0,0 +1,122 @@
+    return new Promise((resolve, reject) => {
+      this._getUserFromDatabase(isEmailLogin, usernameOrEmail)
+        .then((user: SocoboUser) => this._validateUser(user))
+        .then((foundUser: SocoboUser) => this._cryptoUtils.comparePasswords(password, foundUser))
+        .then((compareResult: ComparePwResult) => this._validateComparePasswords(compareResult.isPasswordMatch, compareResult.user))
+        .then((user: SocoboUser) => resolve(this._createLoginResult(user)))
+        .catch((error: any) => reject(error));
+    });
+  }
+
+  private _getUserFromDatabase (isEmailLogin: boolean, usernameOrEmail: string): Promise<SocoboUser> {
+    if (isEmailLogin) {
+      return this._userService.getUserByEmail(usernameOrEmail); 
+    }
+    return this._userService.getUserByUsername(usernameOrEmail);
+  }
+
+  private _validateUser (user: SocoboUser): Promise<SocoboUser> {
+    return new Promise((resolve, reject) => {
+      if (!user) {
+        return reject(new Error("Authentication failed. User not found."));
+      }
+      resolve(user);
+    });
+  }
+
+  private _validateComparePasswords (compareSuccessful: boolean, foundUser: SocoboUser): Promise<SocoboUser> {
+    return new Promise((resolve, reject) => {
+      if (!compareSuccessful) {
+        return reject(new Error("Authentication failed. Wrong password."));
+      }
+      resolve(foundUser);
+    });
+  }
+
+  private _createLoginResult (foundUser: SocoboUser): Promise<LoginResult> {
+    return new Promise((resolve, reject) => {
+      jwt.sign(foundUser, (process.env.TOKEN_SECRET || Config.TOKEN_SECRET), {
+        expiresIn: (process.env.TOKEN_EXPIRATION || Config.TOKEN_EXPIRATION)
+      }, (err, token) => {
+        if (err) {
+          return reject(new Error(`Authentication failed. Error message: ${err.message}.`));
+        }
+        delete foundUser.password;
+        resolve(new LoginResult(token, foundUser));
+      });
+    });
+  }
+
+  register (isEmailLogin: boolean, usernameOrEmail: string, password: string): Promise<SocoboUser> {
+    return new Promise((resolve, reject) => {
+      this._getUserFromDatabase(isEmailLogin, usernameOrEmail)
+        .then((user: SocoboUser) => this._checkIfUserIsAlreadyRegistered(user))
+        .catch((error: any) => {
+          if (ErrorUtils.notFound(error)) {
+            this._cryptoUtils.hashPassword(password)
+              .then((hashedPassword: string) => this._createNewUser(hashedPassword, usernameOrEmail))
+              .then((createdUser: SocoboUser) => resolve(this._returnSavedUser(createdUser)))
+              .catch((error: any) => reject(error))
+          } else {
+            reject(error);
+          }
+        });
+    });
+  }
+
+  private _checkIfUserIsAlreadyRegistered (user: SocoboUser): Promise<any> {
+    return new Promise((resolve, reject) => {
+      if (user) {
+        return reject(new Error("Email or Username is already registered. Please use another one."));
```

Comment:
Yes, I'm with you. We need an error object with message and error code (BaseError) and before we return the error to the user / outside we check the error code and return the right status code and api error etc. We should add this to another issue. This is big enough.